extension/js/common/api/email-provider/gmail/google.ts

```diff
@@ -17,69 +18,73 @@ export class Google {
 
   public static gmailCall = async <RT>(
     acctEmail: string,
-    method: ReqMethod,
     path: string,
-    params: Dict<Serializable> | string | undefined,
-    progress?: ProgressCbs | ProgressDestFrame,
-    contentType?: string
+    params?:
+      | {
+          method: 'POST' | 'PUT';
+          data: Dict<Serializable>;
+          dataType?: 'JSON';
+        }
+      | {
+          method: 'POST';
+          data: string;
+          contentType: string;
+          dataType: 'TEXT';
+        }
+      | {
+          method: 'GET';
+          data?: UrlParams;
+        }
+      | { method: 'DELETE' },
+    progress?: ProgressCbs
   ): Promise<RT> => {
     progress = progress || {};
-    let data, url;
-    if ('upload' in progress) {
+    let url;
+    let dataPart:
+      | { method: 'POST' | 'PUT'; data: Dict<Serializable>; dataType: 'JSON' }
+      | { method: 'POST'; data: string; contentType: string; dataType: 'TEXT' }
+      | { method: 'GET'; data?: UrlParams }
+      | { method: 'DELETE' };
```

Comment:
How about we declear types like below so that we can make code easy to read and simple and avoid code duplication?

```ts
type JsonRequestParams = {
  method: 'POST' | 'PUT';
  data: Dict<Serializable>;
};

type TextRequestParams = {
  method: 'POST';
  data: string;
  contentType: string;
};

type GetRequestParams = {
  method: 'GET';
  data?: UrlParams;
};

type DeleteRequestParams = {
  method: 'DELETE';
};

params?: JsonRequestParams | TextRequestParams | GetRequestParams | DeleteRequestParams
```

Reply:
It'll be a nice improvement, I tried to create separate types this way but looks like it'll require too many changes. For example in [apiCall](https://github.com/FlowCrypt/flowcrypt-browser/blob/issue-1509-fetch/extension/js/common/api/shared/api.ts#L328) in `values` parameter we use `fmt` property, which then is converted to `dataType` in `dataPart` object. Also for `apiCall` method `method` parameter is optional, but it's required for [gmailCall](https://github.com/FlowCrypt/flowcrypt-browser/blob/issue-1509-fetch/extension/js/common/api/email-provider/gmail/google.ts#L19). So we'll need different types with optional and required properties, I think we can leave it as is in the current PR  and file another issue for using such types.

Reply:
Okay, got it